keyboards/unicomp/mini_m/minirazz/matrix.c

```diff
@@ -0,0 +1,159 @@
+/* Copyright 2023 Purdea Andrei
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "quantum.h"
+#include <hardware/spi.h>
+#include <hardware/gpio.h>
+#include <hardware/structs/pads_qspi.h>
+
+matrix_row_t previous_matrix[MATRIX_ROWS];
+
+static inline void setPinInputHigh_atomic(pin_t pin) {
+    ATOMIC_BLOCK_FORCEON {
+        setPinInputHigh(pin);
+    }
+}
+
+static inline void setPinOutput_writeLow(pin_t pin) {
+    ATOMIC_BLOCK_FORCEON {
+        setPinOutput(pin);
+        writePinLow(pin);
+    }
+}
+
+void matrix_init_custom(void) {
+    for (int i=0; i<=7; i++)
+    {
+        setPinInputHigh(i);
+    }
+    for (int i=8; i<=19; i++)
+    {
+        writePinLow(i);
+        setPinInputHigh(i);
+    }
+    spi_init(spi1, 5000000);
+    spi_cpol_t cccpol = 1;
+    spi_cpha_t cccpha = 0;
+    spi_set_format(spi1, 8, cccpol, cccpha, SPI_MSB_FIRST);
+    gpio_set_function(26, GPIO_FUNC_SPI); // SCK
+    gpio_set_function(24, GPIO_FUNC_SPI); // RX (MISO)
+    writePinLow(27);
+    setPinOutput(27);
+
+    for (int gpio=0;gpio<NUM_BANK0_GPIOS;gpio++) {
+        gpio_set_drive_strength(gpio, GPIO_DRIVE_STRENGTH_2MA);
+    }
+
+    for (int gpio=0;gpio<NUM_QSPI_GPIOS;gpio++) {
+        pads_qspi_hw->io[gpio] &= ~((3 << 4) | (1 << 0));
+    }
```

Comment:
This sets drive strength to 2mA and slew rate to 0 (slow).
Constants do exist in `pico-sdk/src/rp2040/hardware_regs/include/hardware/regs/pads_qspi.h`
however these constants are defined separately for each io, even though they are the same.
So using one of the constants on all of them would look like:
```
for (int gpio=0;gpio<NUM_QSPI_GPIOS;gpio++) {
    pads_qspi_hw->io[gpio] &= ~(PADS_QSPI_GPIO_QSPI_SCLK_DRIVE_BITS | PADS_QSPI_GPIO_QSPI_SCLK_SLEWFAST_BITS);
}
```
Or using each definition which it's own register would look loop-unrolled like:
```
pads_qspi_hw->io[0] &= ~(PADS_QSPI_GPIO_QSPI_SCLK_DRIVE_BITS | PADS_QSPI_GPIO_QSPI_SCLK_SLEWFAST_BITS);
pads_qspi_hw->io[1] &= ~(PADS_QSPI_GPIO_QSPI_SD0_DRIVE_BITS | PADS_QSPI_GPIO_QSPI_SD0_SLEWFAST_BITS);
pads_qspi_hw->io[2] &= ~(PADS_QSPI_GPIO_QSPI_SD1_DRIVE_BITS | PADS_QSPI_GPIO_QSPI_SD1_SLEWFAST_BITS);
pads_qspi_hw->io[3] &= ~(PADS_QSPI_GPIO_QSPI_SD2_DRIVE_BITS | PADS_QSPI_GPIO_QSPI_SD2_SLEWFAST_BITS);
pads_qspi_hw->io[4] &= ~(PADS_QSPI_GPIO_QSPI_SD3_DRIVE_BITS | PADS_QSPI_GPIO_QSPI_SD3_SLEWFAST_BITS);
pads_qspi_hw->io[5] &= ~(PADS_QSPI_GPIO_QSPI_SS_DRIVE_BITS | PADS_QSPI_GPIO_QSPI_SS_SLEWFAST_BITS);
```
And the index of the `io[]` array doesn't have any usable defines. Closest to it would be something like `PADS_QSPI_GPIO_QSPI_SCLK_OFFSET`, but that defines of memory address offset from the start of the block, not the index.

Reply:
@sigprof  Do you want me to make one of the above changes? Or perhaps alternatively to define my own constants for these?

Reply:
@sigprof 
> Are these changes actually needed and safe? 

This setting has worked for me without issue on the minirazz, but also on the pico devboard.

Generally speaking one would always want to use the slowest slew rate and drive strength they can get away with. A very fast GPIO output without proper impedance matching will often generate ringing on the bus. This is bad from an EMI perspective, but it is also unsafe because an undershoot below 0V, or an overshoot above the IO voltage can potentially degrade the lifetime of the IOs.

Nobody's touching the SSI BAUDR register beyond the 2nd stage bootloader, so the `PICO_FLASH_SPI_CLKDIV = 2` that's in the board definition files doesn't take effect.

When it comes to hold-time/setup-time timings we're okay.
The rp2040 launches data on the falling edge of the clock, while the spi flash samples it on the rising edge.
The spi flash launches data when it sees the falling edge of the clock (the comment in the 2nd stage bootloader is incorrect, at least when it comes to Winbond chips. the winbond datasheet clearly shows this, see the tCLQV timing)
The rp2040 samples the data on the rising edge of the clock + rx sample delay, which is set to "1", meaning a quarter clock cycle with the current BAUDR. But we'd probably be okay even if rx sample delay wasn't set to 1. So we are quite comfortable with setup and hold times.

For the SPI device the clock rise time[sic, should be rise slew rate, but the unit clears this up] is minimum 0.1V/ns.
That's quite okay as well, at 3.3V at this rise time it means that if you can toggle the clock at `0.1 / 3.3 / 2 * 1000` = 15.15 MHz, then your slew rate is above the minimum. (And we're at 31 MHz, so I'm pretty sure we are okay without checking on a scope)

Maybe the higher drive strength and higher slew rate could help on some boards where the SPI flash is far away from the rp2040, and the capacitance is a lot higher, and/or when also trying to run the SPI flash at a higher frequency. But in this case it's not necessary.

> datasheet mentions the need of the slow slew rate mode is the I2C section (but apparently the default values of I2C1_SCL_PAL_MODE and I2C1_SDA_PAL_MODE have that part wrong).

I would agree with that even in the absence of the note in the datasheet. The rp2040 only supports i2c modes up to 1MHz , at that speed you don't really need fast slew rate. _Maybe_ you can make use of a stronger drive strength, if there's a lot of bus capacitance, because the same lines connect to many devices, and a very strong pull-up resistor could have been chosen, and you may need to fight that, but then keep in mind that all devices you talk to will need that higher drive strength too.

Reply:
DONE, moved into minirazz.c, board_init()